From 741ba56acdb3a02cde541d5a99717d9368fcc58e Mon Sep 17 00:00:00 2001
From: Michal Schmidt <mschmidt@redhat.com>
Date: Mon, 16 Apr 2012 23:27:22 +0200
Subject: [PATCH] [F16] cryptsetup: workaround missing watch rules for dm
 devices

After running mkswap on the decrypted device we never received the
notification about the swap device being ready, so we never executed the
swapon.

The problem is that dm-* devices do not use udev watches in F16.
Workaround it by producing a change uevent manually after mkswap.

device-mapper added the watch rule to 13-dm-disk.rules in F17.

https://bugzilla.redhat.com/show_bug.cgi?id=711394
---
 Makefile.am                   |    3 +++
 src/cryptsetup-generator.c    |   13 ++++++++++---
 src/systemd-cryptsetup-uevent |    7 +++++++
 3 files changed, 20 insertions(+), 3 deletions(-)
 create mode 100755 src/systemd-cryptsetup-uevent

diff --git a/Makefile.am b/Makefile.am
index 2f5dcdb..def8d01 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -202,6 +202,9 @@ if HAVE_LIBCRYPTSETUP
 rootlibexec_PROGRAMS += \
 	systemd-cryptsetup
 
+rootlibexec_SCRIPTS = \
+	src/systemd-cryptsetup-uevent
+
 systemgenerator_PROGRAMS += \
 	systemd-cryptsetup-generator
 endif
diff --git a/src/cryptsetup-generator.c b/src/cryptsetup-generator.c
index a48b7a4..dc1e094 100644
--- a/src/cryptsetup-generator.c
+++ b/src/cryptsetup-generator.c
@@ -67,13 +67,15 @@ static int create_disk(
         char *p = NULL, *n = NULL, *d = NULL, *u = NULL, *from = NULL, *to = NULL, *e = NULL;
         int r;
         FILE *f = NULL;
-        bool noauto, nofail;
+        bool noauto, nofail, tmp, swap;
 
         assert(name);
         assert(device);
 
         noauto = has_option(options, "noauto");
         nofail = has_option(options, "nofail");
+        tmp    = has_option(options, "tmp");
+        swap   = has_option(options, "swap");
 
         if (!(n = unit_name_build_escape("cryptsetup", name, ".service"))) {
                 r = -ENOMEM;
@@ -138,16 +140,21 @@ static int create_disk(
                 name, u, strempty(password), strempty(options),
                 name);
 
-        if (has_option(options, "tmp"))
+        if (tmp)
                 fprintf(f,
                         "ExecStartPost=/sbin/mke2fs '/dev/mapper/%s'\n",
                         name);
 
-        if (has_option(options, "swap"))
+        if (swap)
                 fprintf(f,
                         "ExecStartPost=/sbin/mkswap '/dev/mapper/%s'\n",
                         name);
 
+        if (tmp || swap)
+                fprintf(f,
+                        "ExecStartPost=-" SYSTEMD_CRYPTSETUP_PATH "-uevent '%s'\n",
+                        name);
+
         fflush(f);
 
         if (ferror(f)) {
diff --git a/src/systemd-cryptsetup-uevent b/src/systemd-cryptsetup-uevent
new file mode 100755
index 0000000..4b28db5
--- /dev/null
+++ b/src/systemd-cryptsetup-uevent
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+read MAJ MIN <<EOF
+$(stat -L -c '%t %T' "/dev/mapper/$1")
+EOF
+
+echo change > "/sys/dev/block/$(printf '%d:%d' 0x$MAJ 0x$MIN)/uevent"
